README.md

```diff
@@ -54,8 +54,47 @@ the CNFs produced can be shown UNSAT instantly.
 
 For `k = 6`, `n = 30`, the CNF takes on the order of 17,000 CPU hours to show UNSAT when parallelized using cube&conquer.
 
-The script in `utils/6hole-cube.c` outputs the list of cubes we used to solve `k = 6`, `n = 30`.
-We will add scripts soon for reproducing the entire parallel computation.
+# Cube and Conquer verification
+
+In order to verify that the resulting formula $F$ is indeed UNSAT, two things ought to be checked:
+
+ 1) **(Tautology proof)** The formula $F$ is UNSAT iff $F \land C_i$ is UNSAT for every $i$.
+ 2) **(UNSAT proof)** For each cube $C_i$, the formula $F \land C_i$ is UNSAT.
+
+The **tautology proof** is implied by checking that $F \land (\bigwedge_i \neg C_i)$ is UNSAT, as per Section 7.3 of [Heule and Scheucher](https://arxiv.org/abs/2403.00737).
+
+As a first step, we will generate the main CNF by running:
+```
+lake exe encode hole 6 30 cube_checking/vars.out > cube_checking/6-30.cnf
+```
+
+Then, inside the `cube_checking` folder, the **tautology proof** can be verified by:
+```
+gcc 6hole-cube.c -o cube_gen
+./cube_gen 0 vars.out > cubes.icnf
+python3 cube_join.py -f 6-30.cnf -c cubes.icnf --tautcheck -o taut_if_unsat.cnf
+cadical taut_if_unsat.cnf tautology_proof.lrat --lrat
+cake_lpr taut_if_unsat.cnf tautology_proof.lrat
+```
+Note that this requires the SAT solver [cadical](https://github.com/arminbiere/cadical) to be in the path, and also the verified checker [cake_lpr](https://github.com/tanyongkiam/cake_lpr). As an end result, one sees:
+
+```
+s VERIFIED UNSAT
+```
+
+For the **UNSAT proof**, given the total amount of computation required to run all cubes is probably unfeasible for any verifier, we provide a simple way to verify any given cube to be UNSAT. To generate the formula + cube number $i$ (assuming the steps for the **tautology proof** have been followed), run:
+```
+./cube_gen <i> vars.out > .tmp_cube_<i>
+python3 cube_join.py -f 6-30.cnf -c .tmp_cube_<i> -o with_cube_<i>.cnf
+cadical with_cube_<i>.cnf proof_cube_<i>.lrat --lrat
+cake_lpr with_cube_<i>.cnf proof_cube_<i>.lrat
+```
+To simplify the process, we have added a bash script `solve_cube.sh`, which can be simply run as (for example, for cube number 42):
```

Comment:
Maybe we should add a script for the tautology proof as well?

Reply:
oh wait there is such a script

Reply:
Oops, the tautology script was accidentally pushed. I will add a corrected version.
The reason why I wanted to make the reviewer go command by command is to see what's happening (?)

Reply:
Yeah. I think it's fine to put it in a script if the script is straightforward.

Reply:
Added this :)